File: patron/src/main/java/com/prolificinteractive/patrons/StringSetPreference.java
package com.prolificinteractive.patrons;

import android.annotation.TargetApi;
import android.content.SharedPreferences;
import android.os.Build;
import java.util.Set;

/**
 * Preference wrapper for {@link Set<String>}.
 */
@TargetApi(Build.VERSION_CODES.HONEYCOMB)
public class StringSetPreference extends BasePreference<Set<String>> {
  public StringSetPreference(final SharedPreferences preferences, final String key) {
    this(preferences, key, null);
  }

  public StringSetPreference(
      final SharedPreferences preferences,
      final String key,
      final Set<String> defaultValue) {
    super(preferences, key, defaultValue);
  }

  @Override public Set<String> get() {
    return preferences.getStringSet(key, defaultValue);
  }

  @Override public void set(final Set<String> value) {
    preferences.edit().putStringSet(key, value).apply();
  }
}

File: patron/src/main/java/com/prolificinteractive/patrons/StringSetPreference.kt
package com.prolificinteractive.patrons

import android.annotation.TargetApi
import android.content.SharedPreferences
import android.os.Build

/**
 * Preference wrapper for [Set] of [String].
 */
@TargetApi(Build.VERSION_CODES.HONEYCOMB)
class StringSetPreference @JvmOverloads constructor(
    preferences: SharedPreferences,
    key: String,
    defaultValue: Set<String>? = null
) : BasePreference<Set<String>?>(preferences, key, defaultValue) {

  override fun get(): Set<String>? = preferences.getStringSet(key, defaultValue)

  override fun set(value: Set<String>?) = when (value) {
    null -> delete()
    else -> preferences.edit().putStringSet(key, value).apply()
  }
}
